src/bsoncxx/lib/bsoncxx/v1/element/view.cpp

```diff
@@ -12,19 +12,337 @@
+    std::uint32_t _keylen = {};
+
+    BSONCXX_PRIVATE_WARNINGS_PUSH();
+    BSONCXX_PRIVATE_WARNINGS_DISABLE(GNU("-Wunused"));
+    unsigned char _padding[_padding_size]; // Reserved.
+    BSONCXX_PRIVATE_WARNINGS_POP();
+
+    bool _is_valid = {}; // Last byte.
+
+   public:
+    impl() = default;
+
+    impl(
+        std::uint8_t const* raw,
+        std::uint32_t length,
+        std::uint32_t offset,
+        std::uint32_t keylen,
+        bool is_valid = true)
+        : _raw{raw}, _length{length}, _offset{offset}, _keylen{keylen}, _is_valid{is_valid} {}
+
+    void check() const;
+
+    std::uint8_t const* raw() const {
+        return _raw;
+    }
+
+    std::uint32_t length() const {
+        return _length;
+    }
+
+    std::uint32_t offset() const {
+        return _offset;
+    }
+
+    std::uint32_t keylen() const {
+        return _keylen;
+    }
+
+    bool is_valid() const {
+        return _raw && _is_valid;
+    }
+
+    impl to_invalid() const {
+        auto ret = *this;
+        ret._is_valid = false;
+        return ret;
+    }
+
+    v1::types::id type_id() const {
+        auto const iter = this->iter();
+        return static_cast<v1::types::id>(bson_iter_type(&iter));
+    }
+
+    v1::types::id type_id_unchecked() const {
+        if (auto const iter_opt = this->iter_unchecked()) {
+            return static_cast<v1::types::id>(bson_iter_type(&*iter_opt));
+        }
+        return v1::types::id{}; // BSON_TYPE_EOD
+    }
+
+    v1::stdx::optional<bson_iter_t> iter_unchecked() const {
+        bson_iter_t iter;
+        if (bson_iter_init_from_data_at_offset(&iter, _raw, _length, _offset, _keylen)) {
+            return iter;
+        }
+        return v1::stdx::nullopt;
+    }
+
+    bson_iter_t iter() const {
+        this->check();
+        if (auto iter_opt = this->iter_unchecked()) {
+            return *iter_opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+
+    v1::types::view type_view() const {
+        this->check();
+        return this->type_view_unchecked();
+    }
+
+    v1::types::view type_view_unchecked() const {
+        if (auto opt = v1::types::view::internal::make(_raw, _length, _offset, _keylen)) {
+            return *opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+
+    v1::types::value type_value() const {
+        this->check();
+        if (auto opt = v1::types::value::internal::make(_raw, _length, _offset, _keylen)) {
+            return *opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+};
+
+void view::impl::check() const {
+    static_assert(is_semiregular<impl>::value, "bsoncxx::v1::element::view::impl must be semiregular");
+    static_assert(is_semitrivial<impl>::value, "bsoncxx::v1::element::view::impl must be semitrivial");
+
+    static_assert(sizeof(view::_storage) >= sizeof(view::impl), "insufficient size");
+    static_assert(alignof(view) >= alignof(view::impl), "insufficient alignment");
```

Comment:
Unfortunately, static assertions for `impl` must be moved outside of the (incomplete) class definition into at least one of its out-of-line member function definitions (to allow access of private data members). This similarly applies to `v1::types::value::impl`, where its out-of-line (non-trivial) destructor definition is used instead.